pytorch_lightning/trainer/connectors/checkpoint_connector.py

```diff
@@ -197,15 +197,12 @@ def restore_loops(self) -> None:
         self.trainer.fit_loop.current_epoch = self._loaded_checkpoint["epoch"]
 
         # crash if max_epochs is lower then the current epoch from the checkpoint
-        if (
-            FitLoop._is_max_limit_enabled(self.trainer.max_epochs)
-            and self.trainer.max_epochs is not None
-            and self.trainer.current_epoch > self.trainer.max_epochs
-        ):
-            raise MisconfigurationException(
-                f"You restored a checkpoint with current_epoch={self.trainer.current_epoch},"
-                f" but you have set Trainer(max_epochs={self.trainer.max_epochs})."
-            )
+        if not (self.trainer.max_epochs is None or self.trainer.max_epochs == -1):
+            if self.trainer.current_epoch > self.trainer.max_epochs:
```

Comment:
nit: merge these ifs into one conditional to avoid a level of nesting

Reply:
Yeah I didn't like this either and ended up doing this to make mypy happy + the auto-formatting with casting was really ugly. I just tried again and it's working (not sure what I was thinking earlier) and will push that